drivers/nutdrv_qx.c

```diff
@@ -1209,6 +1216,68 @@ static int	fuji_command(const char *cmd, char *buf, size_t buflen)
 	return (int)strlen(buf);
 }
 
+/* Phoenixtec (Masterguard) communication subdriver */
+static int	phoenixtec_command(const char *cmd, char *buf, size_t buflen)
+{
+	int ret;
+	char *p, *e;
+	char *vis;
+	char *l[] = { "T", "TL", "S", "C", "CT", "M", "N", "O", "SRC", "FCLR", "SS", "TUD", "SSN", NULL }; /* commands that don't return an answer */
+	char **lp;
+
+	if (nut_debug_level >= 4) {
+		stravis(&vis, cmd, VIS_NL | VIS_CSTYLE);
+		upsdebugx(4, "> %s", vis);
+		free(vis);
+	}
+	if ((ret = usb_control_msg(udev, USB_ENDPOINT_OUT | USB_TYPE_VENDOR | USB_RECIP_ENDPOINT, 0x0d, 0, 0, (char *)cmd, strlen(cmd), 1000)) <= 0) {
+		upsdebugx(3, "send: %s (%d)", ret ? usb_strerror() : "timeout", ret);
+		*buf = '\0';
+		return ret;
+	}
+
+	for (lp = l; *lp != NULL; lp++) {
+		const char *q;
+		int b;
+
+		p = *lp; q = cmd; b = 1;
+		while (*p != '\0') {
+			if (*p++ != *q++) {
+				b = 0;
+				break;
+			}
+		}
+		if (b && *q >= 'A' && *q <= 'Z') b = 0; /* "M" not to match "MSO" */
+		if (b) {
+			upsdebugx(4, "command %s returns no answer", *lp);
+			*buf = '\0';
+			return 0;
+		}
+	}
+
+	for (p = buf; p < buf + buflen; p += ret) {
+		if ((ret = usb_interrupt_read(udev, USB_ENDPOINT_IN | 1, p, buf + buflen - p, 1000)) <= 0) {
+			upsdebugx(3, "read: %s (%d)", ret ? usb_strerror() : "timeout", ret);
+			*buf = '\0';
+			return ret;
+		}
+		if ((e = memchr(p, '\r', ret)) != NULL) break;
+	}
+	if (e != NULL && ++e < buf + buflen) {
+		*e = '\0';
+		if (nut_debug_level >= 4) {
+			stravis(&vis, buf, VIS_NL | VIS_CSTYLE);
+			upsdebugx(4, "< %s", vis);
```

Comment:
Already commented about the use of `vis.h` and `stravis()` - I think these can be dropped for many benefits. Also (and especially then) the check for `if (nut_debug_level >= 4) {...}` also becomes redundant since `upsdebugx(4...)` already checks for the requested level.

Same for the case ~50 lines above.